storage/store_pool.go

```diff
@@ -0,0 +1,304 @@
+func (sp *StorePool) start(stopper *stop.Stopper) {
+	stopper.RunWorker(func() {
+		for {
+			var timeout time.Duration
+			sp.mu.RLock()
+			detail := sp.queue.peek()
+			if detail == nil {
+				// No stores yet, wait the full timeout.
+				timeout = sp.timeUntilStoreDead
+			} else {
+				nextAwakeTime := detail.lastUpdatedTime.Add(sp.timeUntilStoreDead)
+				timeout = nextAwakeTime.Sub(time.Now())
+			}
+			sp.mu.RUnlock()
+			select {
+			case <-time.After(timeout):
+				sp.mu.Lock()
+				nextDetail := sp.queue.peek()
+				for nextDetail != nil {
+					deadAsOf := nextDetail.lastUpdatedTime.Add(sp.timeUntilStoreDead)
+					now := time.Now()
+					if now.After(deadAsOf) {
+						detail := sp.queue.dequeue()
+						detail.markDead(now)
+					}
+					nextDetail = sp.queue.peek()
+				}
+				sp.mu.Unlock()
+			case <-stopper.ShouldStop():
+				return
+			}
+		}
+	})
+}
+
+// GetStoreDescriptor returns the latest store descriptor for the given
+// storeID.
+func (sp *StorePool) GetStoreDescriptor(storeID proto.StoreID) *proto.StoreDescriptor {
+	sp.mu.RLock()
+	defer sp.mu.RUnlock()
+
+	detail, ok := sp.stores[storeID]
+	if !ok {
+		return nil
+	}
+	return &detail.desc
+}
+
+// stat provides a running sample size and mean.
+type stat struct {
+	n, mean float64
+}
+
+// Update adds the specified value to the stat, augmenting the sample
+// size & mean.
+func (s *stat) Update(x float64) {
+	s.n++
+	s.mean += (x - s.mean) / s.n
+}
+
+// StoreList holds a list of store descriptors and associated count and used
+// stats for those stores.
+type StoreList struct {
+	stores      []*proto.StoreDescriptor
+	count, used stat
+}
+
+// add includes the store descriptor to the list of stores and updates
+// maintained statistics.
+func (sl *StoreList) add(s *proto.StoreDescriptor) {
+	sl.stores = append(sl.stores, s)
+	sl.count.Update(float64(s.Capacity.RangeCount))
+	sl.used.Update(s.Capacity.FractionUsed())
+}
+
+// GetStoreList returns a storeList that contains all active stores that
+// contain the required attributes and their associated stats.
+// TODO(embark, spencer): consider using a reverse index map from
+// Attr->stores, for efficiency. Ensure that entries in this map still
+// have an opportunity to be garbage collected.
+func (sp *StorePool) GetStoreList(required proto.Attributes, deterministic bool) *StoreList {
+	sp.mu.RLock()
+	defer sp.mu.RUnlock()
+
+	sl := new(StoreList)
+	if !deterministic {
+		for _, detail := range sp.stores {
+			if !detail.dead && required.IsSubset(*detail.desc.CombinedAttrs()) {
+				sl.add(&detail.desc)
+			}
+		}
+		return sl
+	}
+
+	// Sort the stores by key if deterministic is requested. This is only for
+	// unit testing.
+	var storeIDs []int
+	for storeID := range sp.stores {
+		storeIDs = append(storeIDs, int(storeID))
+	}
+	sort.Ints(storeIDs)
+	for storeID := range storeIDs {
+		detail := sp.stores[proto.StoreID(storeID)]
+		if !detail.dead && required.IsSubset(*detail.desc.CombinedAttrs()) {
```

Comment:
We should minimize the duplication between the testing and non-testing branches here. I think it would be best to follow this latter path in all cases, and let the `deterministic` flag only control whether or not we do the sort.